### PR TITLE
improvement: automate changelog management

### PR DESCRIPTION
**Pull request by barryib**
# PR o'clock

## Description

This is a first draft to get feedbacks. There are still some questions about:

- How to manage old changelog ? Put it in the [template](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/786/files#diff-ecd27abc8e9e7087f5375466e0dcc781) ?
- How can I skip new release commits (or "update changelog") in the changelog ? Use `skip/changelog` commit prefix ?
- Ho to ensure that PR titles follow gidelines ? Use something like https://github.com/apps/semantic-pull-requests ?

cc @max-rocket-internet @antonbabenko @js-timbirkett

Resolves https://github.com/terraform-aws-modules/terraform-aws-eks/issues/781

### Todo

- [x] Manage old changelog (Should be added in template ?)
- [x] Avoid adding "Update changelog" or new release commits in the changelog
- [x] Tests to ensure that no one modify the changelog in PRs (don't know yet how to comment back to the PR)
- [x] Use https://github.com/apps/semantic-pull-requests (config ok, have to add the github app)
- [x] Update contribution guide

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+SEMTAG="tools/semtag"
+TAG_QUERY="v10.0.1.."
+scope ?= "minor"
+
+changelog-unrelease:
+	git-chglog -o CHANGELOG.md $(TAG_QUERY)
+
+changelog:
+	git-chglog -o CHANGELOG.md --next-tag `$(SEMTAG) final -s $(scope) -o -f` $(TAG_QUERY)
+
```
Comment (antonbabenko) [inline Makefile]: I would like to have the `release` target also in Makefile:
```
release:
	$(SEMTAG) final -s $(scope)
```

File: tools/semtag [hunk 2]
```diff
@@ -0,0 +1,625 @@
+#!/usr/bin/env bash
+
+PROG=semtag
```
Comment (antonbabenko) [inline tools/semtag]: Probably worth mentioning the source repo we took this script from.

File: Makefile (hunk 1 above)
Comment (barryib) [reply Makefile]: Sure. Added https://github.com/terraform-aws-modules/terraform-aws-eks/pull/786/commits/6f910751246b9a25d6d3ccd16bf1961fa73ee1c6

File: tools/semtag (hunk 2 above)
Comment (barryib) [reply tools/semtag]: Added https://github.com/terraform-aws-modules/terraform-aws-eks/pull/786/commits/6f910751246b9a25d6d3ccd16bf1961fa73ee1c6

**Comment by barryib** [comment]
@antonbabenko @js-timbirkett With your experience with git-chglog, do you have any advice for these questions ?

> - How to manage old changelog ? Put it in the template ?
> - How can I skip new release commits (or "update changelog") in the changelog ? Use skip/changelog commit prefix ?
> - Ho to ensure that PR titles follow gidelines ? Use something like https://github.com/apps/semantic-pull-requests ?

**Comment by antonbabenko** [comment]
I don't have a lot of advanced experience with `git-chglog` but here are my thoughts:
1. I would put the current changelog into `changelog-before-10.md` or something like that, and use the tool for the whole changelog creation with the default template.
2. Maybe we can amend git commits, but I am not sure how to do this for various scenarios. I agree that it does not look nice to have that one commit every time, but I don't know a good solution.
3. We can enable that GitHub app to give it a try and see what happens.

**Comment by barryib** [comment]
> How to manage old changelog ? Put it in the template ?

I've wrote a small makefile which generate changelog with `git-chglog` and replace `<!-- BEGIN GIT-CHGLOG --><!-- END GIT-CHGLOG -->` in the changelog file.

**Comment by barryib** [comment]
@max-rocket-internet @antonbabenko So far here are my feebacks and propositions: 

> - How to manage old changelog ? Put it in the [template](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/786/files#diff-ecd27abc8e9e7087f5375466e0dcc781) ?

I use placeholders in the CHANGELOG.md file and replace it with the generated changelog from git tag v10.0.0

> - How can I skip new release commits (or "update changelog") in the changelog ? Use `skip/changelog` commit prefix ?

Using `skip` as PR or commit prefix and I don't catch them in [git-chglog configuration](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/786/files#diff-fae9ccc8ee18e592aa3cb2f7e08aebcaR8-R19).

> - Ho to ensure that PR titles follow gidelines ? Use something like https://github.com/apps/semantic-pull-requests ?

I'll try this bot during the weekend.

File: .chglog/CHANGELOG.tpl.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+{{ if .Versions -}}
+<a name="unreleased"></a>
+## [Unreleased]
+{{ if .Unreleased.CommitGroups -}}
+{{ range .Unreleased.CommitGroups -}}
+{{ .Title }}
```
Comment (max-rocket-internet) [inline .chglog/CHANGELOG.tpl.md]: Just a small change makes it clearer this is a list

```suggestion
{{ .Title }}:
```

File: .chglog/CHANGELOG.tpl.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{{ end }}
+{{ end -}}
+{{ end -}}
+
+{{ range .Versions }}
+<a name="{{ .Tag.Name }}"></a>
+## {{ if .Tag.Previous }}[{{ .Tag.Name }}]{{ else }}{{ .Tag.Name }}{{ end }} - {{ datetime "2006-01-02" .Tag.Date }}
+{{ if .CommitGroups -}}
+{{ range .CommitGroups -}}
+{{ .Title }}
```
Comment (max-rocket-internet) [inline .chglog/CHANGELOG.tpl.md]: Just a small change makes it clearer this is a list

```suggestion
{{ .Title }}:
```

File: .github/workflows/changelog-check.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  pull_request:
+    paths:
+      - CHANGELOG.md
+
+jobs:
+  changelog-check:
+    name: CHANGELOG Check
+    runs-on: ubuntu-latest
+    steps:
+      # - name: Comment PR
```
Comment (max-rocket-internet) [inline .github/workflows/changelog-check.yaml]: Do we need this whole commented section? Maybe just remove it?

File: .github/semantic.yml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+types:
+  - feat
+  - fix
+  - improvement
+  - docs
+  - refactor
+  - test
+  - ci
+  - chore
+  - skip/changelog
```
Comment (antonbabenko) [inline .github/semantic.yml]: Let's also skip standard prefixes for skipping CI - `[ci skip]`, `[skip ci]`. I am not sure if we have any of those already.

File: .github/workflows/changelog-check.yaml (hunk 5 above)
Comment (barryib) [reply .github/workflows/changelog-check.yaml]: It was for tests. I couldn't make it work. So I removed it for now. I'll try it later.

File: .github/semantic.yml (hunk 6 above)
Comment (barryib) [reply .github/semantic.yml]: I added `skip ci` and `skip changelog` sementics.

**Comment by barryib** [comment]
@max-rocket-internet  After some discussions with @antonbabenko, we decided to move the old changelog into a different file. So the CHANGELOG.md will contain every release notes from v11.0.0 and will be generated from git commits. Are you ok with that ?

**Pull request merged**